src/core/provider/MCAccountValidator.cpp

```diff
@@ -0,0 +1,421 @@
+//
+//  MCAccountValidator.cpp
+//  mailcore2
+//
+//  Created by Christopher Hockley on 22/01/15.
+//  Copyright (c) 2015 MailCore. All rights reserved.
+//
+
+#include "MCAccountValidator.h"
+#include "MCMailProvider.h"
+#include "MCMailProvidersManager.h"
+#include "MCIMAPAsyncSession.h"
+#include "MCPOPAsyncSession.h"
+#include "MCSMTPAsyncSession.h"
+#include "MCNetService.h"
+#include "MCAddress.h"
+#include "MCIMAPOperation.h"
+#include "MCPOPOperation.h"
+#include "MCSMTPOperation.h"
+#include "MCMXRecordResolverOperation.h"
+
+using namespace mailcore;
+
+/* this is the service being tested */
+
+enum {
+    SERVICE_IMAP,  /* IMAP service */
+    SERVICE_POP,   /* POP  service */
+    SERVICE_SMTP,  /* SMTP service */
+};
+
+void AccountValidator::init()
+{
+    mEmail = NULL;
+    mUsername = NULL;
+    mPassword = NULL;
+    mOAuth2Token = NULL;
+    
+    mImapServices = new Array();
+    mSmtpServices = new Array();
+    mPopServices = new Array();
+    
+    mIdentifier = NULL;
+    mImapServer = NULL;
+    mPopServer = NULL;
+    mSmtpServer = NULL;
+    mImapError = ErrorNone;
+    mPopError = ErrorNone;
+    mSmtpError = ErrorNone;
+    
+    mCurrentServiceIndex = 0;
+    mCurrentServiceTested = 0;
+    
+    mProvider = NULL;
+    
+    mOperation = NULL;
+    mQueue = NULL;
+    mResolveMX = NULL;
+    
+    mImapSession = NULL;
+    mPopSession = NULL;
+    mSmtpSession = NULL;
+}
+
+AccountValidator::AccountValidator()
+{
+    init();
+}
+
+AccountValidator::~AccountValidator()
+{
+    MC_SAFE_RELEASE(mEmail);
+    MC_SAFE_RELEASE(mUsername);
+    MC_SAFE_RELEASE(mPassword);
+    MC_SAFE_RELEASE(mOAuth2Token);
+    MC_SAFE_RELEASE(mImapServices);
+    MC_SAFE_RELEASE(mSmtpServices);
+    MC_SAFE_RELEASE(mPopServices);
+    MC_SAFE_RELEASE(mIdentifier);
+    MC_SAFE_RELEASE(mProvider);
+    MC_SAFE_RELEASE(mOperation);
+    MC_SAFE_RELEASE(mQueue);
+    MC_SAFE_RELEASE(mResolveMX);
+    MC_SAFE_RELEASE(mImapSession);
+    MC_SAFE_RELEASE(mPopSession);
+    MC_SAFE_RELEASE(mSmtpSession);
+}
+
+void AccountValidator::start()
+{
+    if (mEmail == NULL) {
+        if (mUsername == NULL) {
+            return;
+        }
+        else {
+            mEmail = mUsername;
+        }
+    }
+    else if (mUsername == NULL){
+        mUsername = mEmail;
+    }
+    
+    mProvider = MailProvidersManager::sharedManager()->providerForEmail(mUsername);
+    
+    if (mProvider == NULL) {
+        resolveMX();
+    }
+    else{
+        startCheckingHosts();
+    }
+}
+
+void AccountValidator::cancel()
+{
+    if(mOperation != NULL)
+        mOperation->cancel();
```

Comment:
I am releasing mOperation, mResolveMX and mQueue below.
mQueue is tested against NULL below too. Is that what you mean? 
Do you mean initialize mQueue in init() rather than later in resolveMX()?